kmir/src/tests/integration/data/prove-rs/unchecked_arithmetic.rs

```diff
@@ -0,0 +1,120 @@
+// @kmir prove-rs: unchecked_add_i32 unchecked_sub_usize unchecked_sub_i16
+
+fn main() {
+    unchecked_add_i32(1,2);
+    unchecked_sub_i8(1, 2);
+    unchecked_sub_u16(1, 2);
+    unchecked_mul_i64(1, 2);
+    unchecked_mul_usize(1, 2);
+}
+
+/// If the precondition is not met, the program is not executed (exits cleanly, ex falso quodlibet)
+macro_rules! precondition {
+    ($pre:expr, $block:expr) => {
+        if $pre { $block }
+    };
+}
+
+fn unchecked_add_i32(a: i32, b: i32) {
+
+    precondition!(
+        ((a as i64 + b as i64) <= i32::MAX as i64) &&
+        ( a as i64 + b as i64  >= i32::MIN as i64),
+        // =========================================
+         unsafe {
+            let result = a.unchecked_add(b);
+            assert!(result as i64 == a as i64 + b as i64)
+        }
+    );
+}
+
+
+macro_rules! unchecked_add_claim_for {
+    ($name:ident, $ty:ident) => {
+        #[allow(unused)]
+        fn $name(a: $ty, b: $ty)
+        {
+            use std::$ty;
+            precondition!(
+                ((a as i128 + b as i128) <= $ty::MAX as i128) &&
+                ( a as i128 + b as i128  >= $ty::MIN as i128),
+                // =========================================
+                 unsafe {
+                    let result = a.unchecked_add(b);
+                    assert!(result as i128 == a as i128 + b as i128)
+                }
+            );
+        }
+    }
+}
+
+unchecked_add_claim_for!(unchecked_add_i8   , i8   );
+unchecked_add_claim_for!(unchecked_add_i16  , i16  );
+// unchecked_add_claim_for!(unchecked_add_i32  , i32  ); // already above
+unchecked_add_claim_for!(unchecked_add_i64  , i64  );
+unchecked_add_claim_for!(unchecked_add_isize, isize);
+unchecked_add_claim_for!(unchecked_add_u8   , u8   );
+unchecked_add_claim_for!(unchecked_add_u16  , u16  );
+unchecked_add_claim_for!(unchecked_add_u32  , u32  );
+unchecked_add_claim_for!(unchecked_add_u64  , u64  );
+unchecked_add_claim_for!(unchecked_add_usize, usize);
```

Comment:
Am I reading this correctly that it's basically making a copy of the above macro for every line here?

I wonder if we could do this using a trait instead? Then make the proof over the trait, and teh prover then goes and discharges the proof obligation for each implementation of the trait (and we define the implementation for the different types we care about).

Just a thought about exploring UX possibilities, no nede to change this PR.

Reply:
Agree it would be great to have a polymorphic test function but there are some technical obstacles if we did that.
1) We would need to call the test function from `main` at the type we want to test, because all code not reachable from `main` gets removed from the MIR.
2) Then again, the monomorphic function to test won't have an easy name, it will be subject to name-mangling in the compiler. That naming is probably relatively stable but not intuitive to the users (therefore not a UX improvement).